Align Python Jacobian to C++, fix #545

gabrielebndn opened the pull request:
Mostly four things:
- Add a benchmark called `timings-jacobian` to compare between `getJointJacobian`, `computeJointJacobians` and related algorithms.
- Use the corresponding C++ function for the Python bindings of `jointJacobian`. Now, the signature and side-effects of `jointJacobian` in Python are completely aligned to C++. The old signature is now deprecated. This fixes #545
- Implemented missing `frameJacobian` function in C++, and use the newly-created function for the Python bindings of `frameJacobian`. The old signature is now deprecated. This also fixes #545
- Change the C++ return policy `getFrameVelocity` and `getFrameAcceleration`. Now, the value is returned instead of being written in a reference

jcarpent commented on src/algorithm/frames.hpp:
```diff
@@ -69,16 +87,61 @@ namespace pinocchio
+                               const typename ModelTpl<Scalar,Options,JointCollectionTpl>::FrameIndex frame_id,
+                               const MotionDense<MotionLike> & frame_v)
+  {
+    const_cast<MotionLike &>(frame_v.derived()) = getFrameVelocity(model, data, frame_id);
```
You don’t have to do a const_cast here

gabrielebndn replied:
The previous implementation had `const_cast`... You mean I should remove `const` the `frame_v` argument? It does seem unnecessary... Or, leave the signature like this and just remove the `const_cast`, would that work?

jcarpent replied:
Yes, it should work without the const_cast.

jcarpent commented on src/algorithm/frames.hpp:
```diff
@@ -69,16 +87,61 @@ namespace pinocchio
+                                   const DataTpl<Scalar,Options,JointCollectionTpl> & data,
+                                   const typename ModelTpl<Scalar,Options,JointCollectionTpl>::FrameIndex frame_id,
+                                   const MotionDense<MotionLike> & frame_a)
+  { const_cast<MotionLike &>(frame_a.derived()) = getFrameAcceleration(model, data, frame_id); }
```
The same apply here

jcarpent commented on benchmark/CMakeLists.txt:
```diff
@@ -75,3 +75,7 @@ ADD_BENCH(timings-eigen)
+
+# timings-jacobian
+#
+ADD_BENCH(timings-jacobian TRUE)
```
Jacobiand are already bench in the main timings file.
Maybe it is not needed to over populate this directory

gabrielebndn replied:
That benchmark only has `computeJointJacobians(model,data,q)` along with many unrelated algorithms. The purpose of this benchmark is specifically comparing the speed of the different ways of computing the Jacobian (mostly, `computeJointJacobians` + `getJointJacobian` vs `jointJacobian`).
It may have meaningful results. For instance, on my machine, it seems `forwardKinematics` + `jointJacobian` is already a bit slower than `computeJointJacobians` (which also computes `forwardKinematics` as a side effect) + `getJointJacobian`. In other words, if, other than the Jacobian, you also need the result of `forwardKinematics`, then it is always better to use `computeJointJacobians`.
I may remove it if you want, but I think it is useful

jcarpent replied:
Yes, you can not compare `forwardKinematics` + `jointJacobian` vs `computeJointJacobians` because this is not the same purpose. In general, we do not need at all `jointJacobian`.
But these results are already known.

jcarpent commented on src/algorithm/frames.hpp:
```diff
@@ -69,16 +87,61 @@ namespace pinocchio
+   */
+  template<typename Scalar, int Options, template<typename,int> class JointCollectionTpl>
+  inline MotionTpl<Scalar, Options>
+  getFrameAcceleration(const ModelTpl<Scalar,Options,JointCollectionTpl> & model,
```
Why did you decide to change this return policy? 
You should keep both I think.

gabrielebndn replied:
Isn't that completely redundant? The return-value-as-argument for this method was my original implementation, but I think it was misguided. In truth, instantiation of a Motion object does not in general dynamically allocate memory, as it only holds a fixed-size 6-element matrix vector. Also, with return-value optimization, I think `getFrameVelocity(model,data,v)` should generate more or less the same code as `v = getFrameVelocity(model,data)`. I see no reason for keeping this signature.

jcarpent commented:
By the way, when you commit, can you add before the commit message which topic you are then considering (like algo, spatial, ...). It is much more readable and this is also the main way of committing on Pinocchio.

gabrielebndn commented:
> By the way, when you commit, can you add before the commit message which topic you are then considering (like algo, spatial, ...). It is much more readable and this is also the main way of committing on Pinocchio.

You are totally right, I did it for the first 3 commits but then I forgot for the rest. I can easily edit the commit messages

jcarpent commented:
@gabrielebndn It will be fine for now concerning the commit messages. I just need you to remove the explicit const_cast.

gabrielebndn commented:
@jcarpent I have really no means to work on it right now (traveling). I will be home this late afternoon, I'll try to fix it then

jcarpent commented:
@gabrielebndn No problem. It was just to summarize the remaining tasks to perform before merging.

jcarpent commented:
@gabrielebndn It would be nice if we can add this feature to the new upcoming release of Pinocchio. Will you have time to deal with it today?

gabrielebndn commented:
I am working on it. I am not on my own computer, so I need to install pretty much everything

gabrielebndn commented:
I have removed the explicit `const_cast` and I have fixed the commit messages. Also, if you want to remove the Jacobian benchmark, you can just discard the last commit

jcarpent commented:
@gabrielebndn Thank you very much for this contribution.
